### PR TITLE
Entities mapping per keyspace

### PR DESCRIPTION
**Pull request by rminchella-deezer**
To allow multi keyspaces and group of tables locked to a specific keyspace, I add a parameter ```orm.entity_managers``` which allow to describe entity folder mapping per entity_manager (using one connection (keyspace)).

Below is the new structure under ```orm``` :

```
cassandra:
    dispatch_events: true
    connections:
        client_vehicle:
              keyspace: "vehicle"
            ...
        client_clothe:
              keyspace: "clothe"
            ...
    orm:
        default_entity_manager: client_vehicle
        entity_managers:
            client_vehicle:
                connection: client_vehicle
                mappings:
                    Boat:
                        dir: "src/Cassandra/BoatEntity"
                    Car:
                        dir: "src/Cassandra/CarEntity"
                    Bike:
                        dir: "src/Cassandra/BikeEntity"
             client_clothe:
                connection: client_clothe
                mappings:
                    Shoe:
                        dir: "src/Cassandra/ShoeEntity"
                    Pant:
                        dir: "src/Cassandra/PantEntity"
                    Shirt:
                        dir: "src/Cassandra/ShirtEntity"
```

That new parameter will allow to crawl the right folder of entities for an entity_manager in the schema create process.

Right now when you do multi connections, all entities are considered and all tables are created in each connection keyspace...

---
This change is [Reviewable](https://reviewable.io/reviews/hendra-huang/cassandrabundle/11)

### CODE REVIEW

**Comment by rminchella-deezer** [comment]
Hello @Hendra-Huang 

If that new parameter is ok for you, I will also update the ReadME.

I thought a lot if it was the right place to define that new parameter.
As that parameter is related to cassandra and orm (entities), I could also put it in a separated config under "orm" as the **mappings** one :

```
->children()
    ->arrayNode('connections_config')
        ->requiresAtLeastOneElement()
        ->useAttributeAsKey('connection')
        ->prototype('array')
            ->treatNullLike([])
            ->performNoDeepMerging()
            ->children()
                ->scalarNode('prefix_entity_folder')->isRequired()->end()
            ->end()
        ->end()
    ->end()
->end()
```

Thank you for your review.

**Comment by IndraGunawan** [comment]
Hi, @rminchella-deezer thanks for your PR
what are your needs so you want to separate entity directory per keyspace?

**Comment by Hendra-Huang** [comment]
Hi, @rminchella-deezer thanks for your PR

CMIIW, you would like to connect to multiple keyspaces within 1 app right ? If that is what you want, I found a good reference [here](https://symfony.com/doc/current/doctrine/multiple_entity_managers.html) . Maybe we can do something similar to this.

Currently our config looks like this.
```
cassandra:
    connections:
        default:
            keyspace: "mykeyspace"
            hosts:
                - 127.0.0.1
                - 127.0.0.2
                - 127.0.0.3
            user: ''
            password: ''
```
To support multiple connection with their own directory, we can introduce a new key called `dir`.

**Comment by rminchella-deezer** [comment]
Hello @Hendra-Huang 

Thank you for your reply.
Using a dir parameter, make sense to follow what is done for entity manager doctrine in SF.
Do you prefer me to put ```dir``` in cassandra.connections.CONNECTION parameter tree ? It's probably better to introduce that parameter per connection in the ```orm``` section as it's related to entity concept. What do you think ?

Thanks for your answer.

**Comment by rminchella-deezer** [comment]
Hello @IndraGunawan 

Yep my need is to allow multi keyspaces for 1 app and so limit a group of entities to one keyspace which is almost represented by a connection in the current implementation.

**Comment by Hendra-Huang** [comment]
@rminchella-deezer yes good idea. Please go ahead with it. 👍

**Comment by rminchella-deezer** [comment]
Hello @Hendra-Huang and @IndraGunawan 

I pushed some updates to handle the ```dir``` parameter in orm.entity_managers configuration.
Below, you can see an example for the structure of the new config parameters :
```
cassandra:
    dispatch_events: true
    connections:
        client_vehicle:
              keyspace: "vehicle"
            ...
        client_clothe:
              keyspace: "clothe"
            ...
    orm:
        default_entity_manager: client_vehicle
        entity_managers:
            client_vehicle:
                connection: client_vehicle
                mappings:
                    Boat:
                        dir: "src/Cassandra/BoatEntity"
                    Car:
                        dir: "src/Cassandra/CarEntity"
                    Bike:
                        dir: "src/Cassandra/BikeEntity"
             client_clothe:
                connection: client_clothe
                mappings:
                    Shoe:
                        dir: "src/Cassandra/ShoeEntity"
                    Pant:
                        dir: "src/Cassandra/PantEntity"
                    Shirt:
                        dir: "src/Cassandra/ShirtEntity"
```

In that example we have 2 connections for 2 keyspaces.
Each connection has his own entity_manager. Each entity_manager is related to some entity folders.

File: DependencyInjection/CassandraExtension.php [hunk 1]
```diff
@@ -34,11 +34,35 @@ public function load(array $configs, ContainerBuilder $container)
...
+    /**
+     * @param string $connectionId
+     * @param string $defaultEmName
+     * @param array  $emConfigs
+     *
+     * @return array
+     */
+    private function getEntityManagerConfiguration($connectionId, $defaultEmName, $emConfigs)
+    {
+        if (!isset($emConfigs[$defaultEmName])) {
```
Comment (Hendra-Huang) [inline DependencyInjection/CassandraExtension.php]: this checking will be executed on every iteration. i think we can move it to outside the loop.

File: DependencyInjection/Configuration.php [hunk 2]
```diff
@@ -104,6 +106,37 @@ private function addOrmSection(ArrayNodeDefinition $rootNode)
...
+                        $v = (array) $v;
+                        // Key that should not be rewritten to the connection config
+                        $excludedKeys = [
+                            'default_entity_manager' => true,
+                            'mappings' => true,
+                            'metadata_cache_driver' => true,
+                        ];
+                        $entityManager = [];
+                        foreach ($v as $key => $value) {
+                            var_dump($key);
```
Comment (Hendra-Huang) [inline DependencyInjection/Configuration.php]: please remove this debugging line

File: README.md [hunk 3]
```diff
@@ -182,6 +182,13 @@ Datacollector is available when the symfony profiler is enabled. The collector a
 
 **NOTE :** The time reported in the data collector may not be the real execution time in case you use the async calls : `executeAsync` and `prepareAsync`
 
+## EntityManager
+
+EntityManager is linked to one connection, so one keyspace in Cassandra.
+In the bundle, you can map some Entity folders to an entityManager to then create some tables (via the SchemaManager) a specific keyspace.
+There is a configuration parameter under ``orm`` called ``entity_managers`` where you can describe each ``entity_manager``.
+The entityManager config contains the linked connection and the entity mapping directories.
```
Comment (Hendra-Huang) [inline README.md:190]: please add 1 more point here
```
If the linked connection can't be found, will fallback to default connection
```

**Review by Hendra-Huang — COMMENTED** [review]
good job man. just left some small comments

File: DependencyInjection/CassandraExtension.php (hunk 1 above)
Comment (rminchella-deezer) [reply DependencyInjection/CassandraExtension.php]: I don't see any issue by doing that way.
I made an update to follow your feedback but I am not sure it's better...

**Comment by rminchella-deezer** [comment]
Hello @Hendra-Huang,
 
Thanks for your feedback.
I did the fixes even if for the defaut entity manager config check, I don't see any issue with my previous code.

**Review by Hendra-Huang — APPROVED** [review]

**Pull request merged**